tools/halide_image_io.h

```diff
@@ -959,8 +959,150 @@ bool find_imageio(const std::string &filename, ImageIO<ImageType, check> *result
     return check(false, err.c_str());
 }
 
+// Given something like ImageType<Foo>, produce typedef ImageType<Bar>
+template<typename ImageType, typename ElemType>
+struct ImageTypeWithElemType {
+    using type = decltype(std::declval<ImageType>().template as<ElemType>());
+};
+
+// Must be constexpr to allow use in case clauses.
+inline constexpr int halide_type_code(halide_type_code_t code, int bits) {
+    return (((int) code) << 8) | bits;
+}
+
 }  // namespace Internal
 
+struct ImageTypeConversion {
+    // Convert an Image from one ElemType to another, where the src and
+    // dst types are statically known (e.g. Buffer<uint8_t> -> Buffer<float>).
+    // You'd normally call this with an explicit type for DstElemType and
+    // allow ImageType to be inferred, e.g.
+    //     Buffer<uint8_t> src = ...;
+    //     Buffer<float> dst = convert_image<float>(src);
+    template<typename DstElemType, typename ImageType>
+    static auto convert_image(const ImageType &src) -> 
+            typename Internal::ImageTypeWithElemType<ImageType, DstElemType>::type {
+        using SrcImageType = ImageType;
+        using SrcElemType = typename SrcImageType::ElemType;
+
+        using DstImageType = typename Internal::ImageTypeWithElemType<ImageType, DstElemType>::type;
+
+        static_assert(SrcImageType::has_static_halide_type, 
+                      "convert_image_s2s() requires a statically-typed image");
+
+        DstImageType dst = DstImageType::make_with_shape_of(src);
+        const auto converter = [](DstElemType &dst_elem, SrcElemType src_elem) {
+            dst_elem = Internal::convert<DstElemType>(src_elem);
+        };
+        // TODO: do we need src.copy_to_host() here?
+        dst.for_each_value(converter, src);
+
+        return dst;
+    }
+
+    // Convert an Image from one ElemType to another, where the src type
+    // is statically known but the dst type is not 
+    // (e.g. Buffer<uint8_t> -> Buffer<>(halide_type_t)). 
+    template <typename DstElemType = void,
+              typename ImageType, 
+              typename std::enable_if<ImageType::has_static_halide_type>::type * = nullptr>
+    static auto convert_image(const ImageType &src, const halide_type_t &dst_type) -> 
+            typename Internal::ImageTypeWithElemType<ImageType, void>::type {
+        // The enable_if ensures this will never fire; this is here primarily
+        // as documentation and a backstop against breakage.
+        static_assert(ImageType::has_static_halide_type, 
+                      "This variant of convert_image() requires a statically-typed image");
+
+        const halide_type_t src_type = src.type();
+        if (src_type == dst_type) {
+            // My, that was easy.
+            return src;
```

Comment:
Hrm... that is a good question. I suppose that making a copy in all cases is probably safer.